### PR TITLE
Simulation from Causal Graph

### LINKED ISSUE #61: [MB] Week of March 22
- [x] Estimate models based on simulated causal graph - Priority 1:
   - Deliverable: pushed updated notebook with estimated models
   - Estimate: 2 hours
   - Actual: 2 hours

- [x] Replace choice simulation code - Priority 2:
   - Deliverable: pushed code to the simulation branches with the new functions.
   - Estimate: 1 hour
   - Actual: 1 hour

- [x] Add inline comments and docstrings where needed - Priority 3:
   - Deliverable: updated pushed notebooks with inline comments and docstrings
   - Estimate: 5 hours
   - Actual: 4 hours

- [x] Add visualizations of estimated parameter distributions - Priority 4:
   - Deliverable: updated pushed notebooks with inline added visualizations
   - Estimate: 2 hours
   - Actual: 2 hours

- [x] Edit independent-variable simulation code to remove outliers, such as by using truncated distributions - Priority 5.
   - Deliverable: Pushed commits with pseudocode of how one plans to make the desired edits.
   - Deliverable: Pushed commits with code change, showing a visualization of the resulting distribution and evidence the outliers are gone.
   - Estimated: 2 hours
   - Actual: 1.5 hours

- [x] Weekly Meeting:
   -  Estimate: 2 hours
   - Actual: 2 hours

### PR DESCRIPTION
**Pull request by bouzaghrane**
I am starting the pull request to review the code for the workflow from last week.

### CODE REVIEW

File: notebooks/working/5.0-mab-simulation-causal-graph.md [hunk 1]
```diff
@@ -0,0 +1,994 @@
...
+
+# Example Application Using Bike Data
+
+
+## Data Ingestion and Exploration
+
+```python
+# Create a variable for the path to the long format data for
+# the multinomial choice model
+PATH = '/Users/mobouzaghrane/Documents/GitHub/tr_b_causal_2020/'\
```
Comment (timothyb0912) [inline notebooks/working/5.0-mab-simulation-causal-graph.md]: Could you replace this with a relative path so that it'll work for other users? 
```suggestion
PATH = '../../'\
```

Comment (timothyb0912) [reply notebooks/working/5.0-mab-simulation-causal-graph.md]:
```suggestion
PATH = '../../'\
```

Comment (bouzaghrane) [reply notebooks/working/5.0-mab-simulation-causal-graph.md]: Will do!

File: notebooks/working/5.0-mab-simulation-causal-graph.md [hunk 2]
```diff
@@ -961,6 +961,115 @@ def FitAlternativeRegression(regressions, reg_types, data):
...
+### Drive Alone 
+
+```python
+V_Drive_Alone.draw()
+```
+
+```python
+Drive_Alone_Df = bike_data_long[bike_data_long['mode_id']==1]
+Drive_Alone_Df.reset_index(drop=True,inplace=True)
+Drive_Alone_Reg = FitAlternativeRegression(regressions={1:('total_travel_distance','total_travel_cost'),
```
Comment (timothyb0912) [inline notebooks/working/5.0-mab-simulation-causal-graph.md:1290]: Small nitpicks around line length for readability.

Also, having the name of the variable by itself for a multi-line variable definition allows you to change the variable name freely while still having easy to understand git-diffs.

```suggestion
Drive_Alone_Reg =\
    FitAlternativeRegression(
        regressions={1:('total_travel_distance','total_travel_cost'),
```

Similar logic applies to putting `FitAlternativeRegression` by itself on a line as it's a multi-line function call.

Comment (bouzaghrane) [reply notebooks/working/5.0-mab-simulation-causal-graph.md:1290]: Ah! got it! Lots to learn from you brother! Thanks so much!

File: notebooks/working/5.0-mab-simulation-causal-graph.md [hunk 3]
```diff
@@ -961,6 +961,115 @@ def FitAlternativeRegression(regressions, reg_types, data):
...
+### Shared-3+
+
+```python
+V_Shared_3p.draw()
+```
+
+```python
+Shared_3p_Df = bike_data_long[bike_data_long['mode_id']==3]
+Shared_3p_Df.reset_index(drop=True,inplace=True)
+Shared_3p_Reg = FitAlternativeRegression(regressions={1:('total_travel_distance','total_travel_cost'),
```
Comment (timothyb0912) [inline notebooks/working/5.0-mab-simulation-causal-graph.md:1320]: You might get to this in the subsequent commits, but am I wrong in thinking that your pseudocode is:

1. get distribution of nodes with no parents.
2. get distribution of nodes with parents (i.e. fit the regressions)
3. simulate nodes with parents
4. simulate availability
5. simulate nodes with no parents

But then your actual code does 1, 2, 4, 5?
Just wanted to make sure I haven't missed anything!

Comment (bouzaghrane) [reply notebooks/working/5.0-mab-simulation-causal-graph.md:1320]: Actually, the pseudocode should do the following:

1. get distribution of nodes with no parents.
2. get distribution of nodes with parents (i.e. fit the regressions)
3. simulate nodes with no parents
4. simulate nodes with parents
5. simulate availability
6. simulate choices.

I simulate nodes with parents between lines 1089-1293

File: notebooks/working/5.0-mab-simulation-causal-graph.md [hunk 4]
```diff
@@ -0,0 +1,1432 @@
...
+
+### Bike
+
+```python
+V_Bike.draw()
+```
+
+No need to simulate any other variables since Bike Total Travel Distance was already simulated
+
+
```
Comment (bouzaghrane) [inline notebooks/working/5.0-mab-simulation-causal-graph.md]: @timothyb0912 here is where I simulate data for nodes with parents.

Comment (timothyb0912) [reply notebooks/working/5.0-mab-simulation-causal-graph.md]: ah got it got it!

I was ging through the commits one by one and was on `98f2b`. I left a comment there but wasn't sure if it would show as a comment on the specific commit (which didn't have this yet) or as a comment on the overall file.

It seems it was the latter!

File: notebooks/working/5.0-mab-simulation-causal-graph.md (hunk 2 above)
Comment (timothyb0912) [reply notebooks/working/5.0-mab-simulation-causal-graph.md:1290]: psh, the learning goes both ways! Thanks for working on this during the craziness going on!

File: notebooks/working/5.0-mab-simulation-causal-graph.md [hunk 5]
```diff
@@ -0,0 +1,1432 @@
...
+        if reg_types[reg] == 'linear':
+            data_x = sm.add_constant(data[regressions[reg][0]])
+            data_y = data[regressions[reg][1]]
+
+            model = sm.OLS(data_y, data_x)
+
+            results = model.fit()
+            regression_results[regressions[reg][1]+'_on_'+regressions[reg][0]] = results
+        
+        if reg_types[reg] == 'logistic':
```
Comment (timothyb0912) [inline notebooks/working/5.0-mab-simulation-causal-graph.md]: Right now we can only do logistic regression (i.e. binary features) but we may want multinomial regression for multi-category features.

File: notebooks/working/5.0-mab-simulation-causal-graph.md [hunk 6]
```diff
@@ -0,0 +1,1432 @@
...
+            data_x = sm.add_constant(data[regressions[reg][0]])
```
Comment (timothyb0912) [inline notebooks/working/5.0-mab-simulation-causal-graph.md]: To note Hassan's point, right now we can only do one variable in the regression but we want to be more general (more than one variable).

**Comment by bouzaghrane** [comment]
@timothyb0912 @hassanobeid1994 take a look at the notebook and run it and check out the results of the estimation.

**Review by timothyb0912 — CHANGES_REQUESTED** [review]
Nice progress on this Amine!
At the end of your current notebook can you do the following?
1. Extract the estimated parameters into a single matrix of shape `(num_simulations, num_parameters)`
2. For each parameter,
   1. plot a histogram and/or kernel density estimate of the simulations
   2. plot the data generating parameter value as a vertical line
   3. label all axes, despine, and show the plot? (see my notebooks for examples.)

This will allow people to visually assess whether our estimation and simulation procedures are working together as expected.

**Comment by bouzaghrane** [comment]
@timothyb0912 Will make those changes! Thanks bro.

**Comment by bouzaghrane** [comment]
@timothyb0912 @hassanobeid1994 check out the plots I just added.

File: notebooks/working/5.0-mab-simulation-causal-graph.md [hunk 7]
```diff
@@ -1840,22 +1885,20 @@ for sim_size, number in zip(simulation_sizes, sim_number):
...
-    sns.distplot(sim_par, kde=False, bins=bins, label='Simulated Parameters')
-    plt.axvline(model.coefs[sim_par.name], color='black', ls='--',label='Data Generating Parameter')
-    plt.title(label='Histogram of '+'"'+sim_par.name+'"', fontdict={'fontsize':16})
-    plt.ylabel('Frequency', rotation=0, labelpad=50)
-    plt.legend()
+# Loop around the different parameters
+# and plot the figures:
+
+for par in sim_estimated_par.columns:
+    PlotParams(sim_estimated_par[par],mnl_model, (10,6))
```
Comment (timothyb0912) [inline notebooks/working/5.0-mab-simulation-causal-graph.md:1899]: Beautiful. These plots are exactly what I was looking for and the helper function to plot them is much appreciated!

**Review by bouzaghrane — COMMENTED** [review]
@timothyb0912 Just made the changes. Let me know if you approve.

**Review by timothyb0912 — APPROVED** [review]

**Review by timothyb0912 — CHANGES_REQUESTED** [review]
Okay, I think this notebook is in great shape.

I think the remaining thing to do is (on each simulation in your for-loop):
- estimate causal effects using the specified causal graph with dependent explanatory variables
- estimate causal effects using a misspecified causal graph with independent explanatory variables
- estimate true causal effects using the data-generating parameters.

Then you can collect all of these various estimates into three dataframes.

Finally, you can make two plots. Each plot will show two distributions, one of which will be the distribution of true causal effects (i.e. those computed with the data-generating parameters).
- One plot will additionally show the distribution of estimated causal effects using a graph with independent explanatory variables.
- Other plot will additionally show the distribution of estimated causal effects using a graph with dependent explanatory variables.

The goal is to see that the plot based on the correct causal graph has high overlap between the distribution of estimated and true causal effects while the plot based on the misspecified causal graph with independent variables shows low overlap.

**Review by timothyb0912 — APPROVED** [review]
